nixos/modules/services/misc/matrix-registration.nix

```diff
@@ -0,0 +1,218 @@
+{ config, pkgs, lib, ... }:
+
+with lib;
+let
+  dataDir = "/var/lib/matrix-registration";
+  cfg = config.services.matrix-registration;
+  format = pkgs.formats.yaml { };
+  matrix-registration-config =
+    format.generate "config.yaml" cfg.settings;
+  matrix-registration-cli-wrapper = pkgs.stdenv.mkDerivation {
+    name = "matrix-registration-cli-wrapper";
+    buildInputs = [ pkgs.makeWrapper ];
+    buildCommand = ''
+      mkdir -p $out/bin
+      makeWrapper ${pkgs.matrix-registration}/bin/matrix-registration "$out/bin/matrix-registration" \
+        --add-flags "--config-path='${matrix-registration-config}'"
+    '';
+  };
+
+in
+{
+  options.services.matrix-registration = {
+    enable = mkEnableOption "Start Matrix Registration";
+
+    settings = mkOption {
+      default = { };
+      description = "matrix-registration configuration";
+      type = types.submodule {
+        freeformType = format.type;
+
+        options = {
+          server_location = mkOption {
+            type = types.str;
+            description = "The client base URL";
+            example = "https://matrix.example.tld";
+          };
+          server_name = mkOption {
+            type = types.str;
+            description = "The server_name";
+            example = "example.tld";
+            default = optionalString config.services.matrix-synapse.enable
+              config.services.matrix-synapse.server_name;
+          };
+          registration_shared_secret = mkOption {
+            type = types.str;
+            description =
```

Comment:
Please mention it is a bad idea to use this because secrets and the nix store. Recommend credentials file and possible provide an example to utilize this as well. Finally, please don't put a default value in - we don't want to make life easier/encourage people to store secrets in the nix store.

Reply:
right makes sense, will change that.
My initial thought was just since afaik this secret ends up in the nix store eitherway, since matrix-synapse doesn't have any other mechanism to set it, it doesn't really matter.

Reply:
I assume the `--secrets-path` argument would allow you to keep secrets out of the configuration file - is that not the case?

Reply:
no that's correct. but this secret needs to also be set for synapse.
they need to be the same for them to communicate.
https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/misc/matrix-synapse.nix#L489

so if you use this and synapse on the same machine it will end up in your nix store, since synapse (afaik) doesn't implement any mechanism to supply this secret securely.

Reply:
actually there is, sorry:
https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/misc/matrix-synapse.nix#L633